include/libtorrent/aux_/portmap.hpp

```diff
@@ -66,6 +66,32 @@ namespace libtorrent { namespace aux {
 	protected:
 		~portmap_callback() {}
 	};
+
+		struct base_mapping
+	{
+		enum class action : std::uint8_t { none, add, del };
+
+		// the time the port mapping will expire
+		time_point expires;
+
+		action act = action::none;
+
+		// the external (on the NAT router) port
+		// for the mapping. This is the port we
+		// should announce to others
+		int external_port = 0;
+
+			portmap_protocol protocol = portmap_protocol::none;
```

Comment:
it looks like the indentation is wrong from here on. At first I thought the ``to_string()`` functions were still members